fennel/expr/expr.py

```diff
@@ -0,0 +1,748 @@
+
+
+class ListLen(ListOp):
+    pass
+
+
+@dataclass
+class ListContains(ListOp):
+    item: Expr
+
+
+@dataclass
+class ListGet(ListOp):
+    index: Expr
+
+
+class ListNoop(ListOp):
+    pass
+
+
+class _List(Expr):
+    pass
+
+
+#######################################################
+
+
+class Literal(Expr):
+    def __init__(self, c: Any, type: Type):
+        super(Literal, self).__init__()
+        if getattr(c.__class__, FENNEL_STRUCT, False):
+            val = json.dumps(c.as_json())
+        else:
+            try:
+                val = json.dumps(c)
+            except TypeError:
+                val = json.dumps(str(c))
+        self.c = val
+        self.dtype = type
+
+
+class Unary(Expr):
+    def __init__(self, op: str, operand: Any):
+        valid = ("~", "len", "str")
+        if op not in valid:
+            raise InvalidExprException(
+                "unary expressions only support %s but given '%s'"
+                % (", ".join(valid), op)
+            )
+        operand = make_expr(operand)
+        if not isinstance(operand, Expr):
+            raise InvalidExprException(
+                "operand can only be an expression but got %s instead" % operand
+            )
+        self.op = op
+        self.operand = operand
+        operand.edge(self)
+        super(Unary, self).__init__()
+
+    def __str__(self) -> str:
+        if self.op in ["len", "str"]:
+            return f"{self.op}({self.operand})"
+        else:
+            return f"{self.op}{self.operand}"
+
+
+class Binary(Expr):
+    def __init__(self, left: Any, op: str, right: Any):
+        valid = (
+            "+",
+            "-",
+            "*",
+            "/",
+            "//",
+            "%",
+            "and",
+            "or",
+            "==",
+            ">=",
+            ">",
+            "<",
+            "<=",
+            "!=",
+            "[]",
+            "in",
+        )
+        if op not in valid:
+            raise InvalidExprException(
+                "binary expressions only support %s but given '%s'"
+                % (", ".join(valid), op)
+            )
+        left = make_expr(left)
+        right = make_expr(right)
+        if not isinstance(left, Expr):
+            raise InvalidExprException(
+                "left can only be an expression but got %s instead" % left
+            )
+        if not isinstance(right, Expr):
+            raise InvalidExprException(
+                "right can only be an expression but got %s instead" % right
+            )
+        self.left = left
+        self.op = op
+        self.right = right
+        left.edge(self)
+        right.edge(self)
+
+        dtype = right.dtype if right.dtype is not None else left.dtype
+        super(Binary, self).__init__(None, dtype)
+
+    def __str__(self) -> str:
+        if self.op == "[]":
+            return f"{self.left}[{self.right}]"
+        else:
+            return f"{self.left} {self.op} {self.right}"
+
+
+class When(Expr):
```

Comment:
When is an expresssion so technically one could write `when(something).list(...)` right? do we have protections in place to ensure that when must be followed by then?